spinalcordtoolbox/deepseg_gm/deepseg_gm.py

```diff
@@ -210,41 +176,21 @@ def segment_volume(ninput_volume, model_name,
     :return: segmented slices.
     """
     gmseg_model_challenge = DataResource('deepseg_gm_models')
-    model_path, metadata_path = model.MODELS[model_name]
-
-    metadata_abs_path = gmseg_model_challenge.get_file_path(metadata_path)
-    with open(metadata_abs_path) as fp:
-        metadata = json.load(fp)
-
-    volume_size = np.array(ninput_volume.shape[0:2])
-    small_input = (volume_size <= SMALL_INPUT_SIZE).any()
-
-    if small_input:
-        # Smaller than the trained net, don't crop
-        net_input_size = volume_size
-    else:
-        # larger sizer, crop at 200x200
-        net_input_size = (SMALL_INPUT_SIZE, SMALL_INPUT_SIZE)
-
-    deepgmseg_model = model.create_model(metadata['filters'],
-                                         net_input_size)
```

Comment:
> * `sct_deepseg_gm` needed extra work: 
>   * The original model spec defined its own input layer on the fly based on the input image size.
>       * If the input image was smaller than 200x200, the input layer would be set to the image size. 
>       * If the input image was larger, then the image would be center-cropped to 200x200.
>   * So, In order to use a fixed-input-size ONNX model, the center-cropping code was replaced with cropping AND padding code to fit the input image to 200x200, regardless of size.

Unfortunately, this change was not backwards-compatible as I had hoped. While the outputs of both models are 99% similar, there are some minute differences between the old and new outputs. For example, using `sct_testing_data/t2s.nii.gz`

* Old: Creating a 96x106 input layer:



* New: Padding the 96x106 image to 200x200, then passing it to the 200x200 input layer:

![image](https://user-images.githubusercontent.com/16181459/159373212-a765a8fd-d108-4297-8188-97a4dcf5a246.png)

Depending on how severe these small deviations are, I may need to try to look into creating a [dynamic input layer](https://stackoverflow.com/questions/62237631/using-dynamic-input-shape-in-keras) for the original Keras model, then converting _that_ to ONNX.

Reply:
> Depending on how severe these small deviations actually are, I may need to try to look into creating a model with a [variable-dimension input layer](https://github.com/keras-team/keras/issues/1920#issuecomment-193883690) for the original Keras model, then try converting _that_ to ONNX.

I've tried this, but I don't believe it is a viable option either, as the Global Pooling layer is dependent on the input size, which would vary for images smaller than 200x200. :(

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/84060e21e95ec75a0b0c2df195ac1d95a8ca5fab/spinalcordtoolbox/deepseg_gm/model.py#L133-L136

Reply:
One other thing to consider: The alternative to using ONNX is to retrain new models from scratch (#3046), which would almost assuredly create _radically_ different results.

In that context, we could consider ourselves lucky to only run into tiny deviations in a specific corner case, while still preserving identical output in all other cases.

Reply:
Aha! I realize I've been zero-padding, as opposed to padding with non-zero values (e.g. reflections, edge-extension, etc.).  This meant that the normalization step was skewed greatly towards the large zero-value areas:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/47922b1a9249828225ae2f6078cca3493979513f/spinalcordtoolbox/deepseg_gm/deepseg_gm.py#L199

When I switch the padding to `symmetric`, then the predicted regions are now much closer to the Keras models for small images:

<details>

* Old method: Creating a 96x106 input layer on the fly (Keras model):
* New method: Symmetry-padding the 96x106 image to 200x200, then passing it to a fixed 200x200 input layer (ONNX model):

</details>

Now there are 0-2 voxel differences across all slices.

Reply:
During our last SCT meeting (2022-03-22), we came to the conclusion that these small deviations are not enough of a concern to block progress on this PR. 

It was noted during the meeting that encountering images <200x200 will likely be quite rare, given the resampling to 0.25mm resolution, and especially as imagining technology continues to advance (producing higher fidelity images).